### ISSUE TITLE
Can't install composer package on OSX Sierra due to OpenSSL dependency warning

### ISSUE THREAD

**Issue by CodeCommander** [Spomky-Labs/jose#148]
I'm trying to install:

> $ ./composer.phar require spomky-labs/jose
> Using version ^6.1 for spomky-labs/jose
> ./composer.json has been updated
> Loading composer repositories with package information
> Updating dependencies (including require-dev)
> Your requirements could not be resolved to an installable set of packages.
> 
>   Problem 1
>     - spomky-labs/jose v6.1.1 requires lib-openssl * -> the requested linked library openssl has the wrong version installed or is missing from your system, make sure to have the extension providing it.
>     - spomky-labs/jose v6.1.0 requires lib-openssl * -> the requested linked library openssl has the wrong version installed or is missing from your system, make sure to have the extension providing it.
>     - Installation request for spomky-labs/jose ^6.1 -> satisfiable by spomky-labs/jose[v6.1.0, v6.1.1].

It looks like there is a dependency on OpenSSL failing?

When I do:

> $./composer.phar show --platform | grep openssl
> ext-openssl         0        The openssl PHP extension (actual version: ) 

Does that mean I don't have OpenSSL extension available?

If I do:

> $ php -i | grep OpenSSL
> OpenSSL support => enabled
> OpenSSL Library Version => LibreSSL 2.2.7
> OpenSSL Header Version => LibreSSL 2.2.7
> Native OpenSSL support => enabled

It looks like the library is available, it is maybe just not the exact version expected?

Is this a composer issue maybe? Or can it be fixed with a change to the package?

**Comment by Spomky** [comment]
Hi, 

Thank you for reporting this issue. 
Can you execute the following command line and verify `lib-openssl` is correctly listed: `composer show -p`

**Comment by Spomky** [comment]
Hi, 

I found an interesting issue reported [on the Composer repository] (https://github.com/composer/composer/issues/5749)  related to missing extensions. 

I am not sure if it could help, but can you try to use the library by using the `--ignore-platform-reqs` parameter and tell me if it works fine or not?

**Comment by moderndeveloperllc** [comment]
@CodeCommander Just curious - what package manager are you using? Could this be related to the fact you are using LibreSSL? What does `composer show -p | grep ssl` output? For reference, I'm using Homebrew and get:
```
$ php -i | grep OpenSSL
SSL Version => OpenSSL/1.0.2j
OpenSSL support => enabled
OpenSSL Library Version => OpenSSL 1.0.2j  26 Sep 2016
OpenSSL Header Version => OpenSSL 1.0.2j  26 Sep 2016
OpenSSL support => enabled
$ composer show -p | grep ssl
ext-openssl         7.1.0RC6  The openssl PHP extension
lib-openssl         1.0.2.10  OpenSSL 1.0.2j  26 Sep 2016
```

**Comment by CodeCommander** [comment]
@moderndeveloperllc  I have macports installed, but I believe my PHP installation is using entirely the standard MacOSX Sierra configuration, I haven't explicitly upgraded it, and I was unable to (easily) install openssl extension manually due to security restrictions imposed by OSX.

Here is the output from those commands:

> $ php -i | grep OpenSSL
> OpenSSL support => enabled
> OpenSSL Library Version => LibreSSL 2.2.7
> OpenSSL Header Version => LibreSSL 2.2.7
> Native OpenSSL support => enabled
> $ ./composer.phar show -p | grep ssl
> ext-openssl         0        The openssl PHP extension (actual version: )

**Comment by CodeCommander** [comment]
@Spomky `./composer.phar require --ignore-platform-reqs spomky-labs/jose` seemed to do the trick to install it. I'll try it out and see if it works. Thanks!

**Comment by CodeCommander** [comment]
Hmm, now I get an error about `Call to undefined function gmp_init()`. It looks like this extension maybe is not enabled on the default Mac OSX PHP install and I need to install it with brew. I guess using this package with a standard Mac PHP configuration is not going to be possible.

**Comment by Spomky** [comment]
I am sorry but as I have no Mac computer I cannot reproduce that issue and find a quick solution.
Nevertheless it seems that you have to recompile PHP with the OpenSSL support (see https://www.farces.com/wikis/naked-server/php/php-openssl/).
For other extensions such as GMP, I found [that answer that could help you](http://stackoverflow.com/questions/33489530/how-to-enable-gmp-in-php-on-os-x) (to be modified according to your system).

**Comment by moderndeveloperllc** [comment]
@CodeCommander There are some packages/extensions out there that have issues with Secure Transport (the `Native OpenSSL support => enabled`) which is what macOS uses natively instead of OpenSSL. For instance [cURL can have problems](http://stackoverflow.com/questions/26461966/osx-10-10-curl-post-to-https-url-gives-sslread-error) in PHP. I've had to move away from the stock installs to using [Homebrew](https://getgrav.org/blog/macos-sierra-apache-multiple-php-versions) as MacPorts can be a touch slow to update things like Apache. This means that your configuration is very similar to what would be on a production Linux server. [Docker](https://docs.docker.com/docker-for-mac/) would be the most complete solution, but it's convoluted to get working just right.

I'm not sure what @Spomky is using that requires specifically that OpenSSL extension, though I'm guessing it has to do with [certificate handling functions](http://php.net/manual/en/ref.openssl.php). If you are really curious, you can try to do an `openssl_encrypt()` in a test script to see if those functions are available to you.

**Comment by Spomky** [comment]
Hi @CodeCommander ,

I confirm you that the issue comes from the `Native OpenSSL support` installed on Sierra.
As @moderndeveloperllc mentioned, I tried to install the change that OpenSSL version. I followed the steps of [this blog post](https://www.farces.com/wikis/naked-server/php/php-openssl/) and it works fine.

Just install Homebrew:

```sh
/usr/bin/ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"
```

Execute the following command lines:

```sh
brew tap homebrew/dupes
brew tap homebrew/versions
brew tap homebrew/php
brew install curl --with-libssh2 --with-openssl
```

And you are done. Execute `php -i | grep "SSL Version"` to verify your OpenSSL version (should be `OpenSSL...` instead of `LibreSSL...`).

@moderndeveloperllc thank you for your support, I really appreciate it.
